File: HouseRobber.cpp
/**
 * Basic DSA - House Robber (LeetCode 198)
 * Max sum of non-adjacent elements.
 */
#include <iostream>
#include <vector>
#include <algorithm>

int rob(std::vector<int>& nums) {
    int prev = 0, cur = 0;
    for (int x : nums) {
        int next = std::max(cur, prev + x);
        prev = cur;
        cur = next;
    }
    return cur;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: House Robber\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {1, 2, 3, 1};
    std::cout << "[1,2,3,1] -> " << rob(a1) << "\n";

    std::vector<int> a2 = {2, 7, 9, 3, 1};
    std::cout << "[2,7,9,3,1] -> " << rob(a2) << "\n";

    std::vector<int> a3 = {2, 1, 1, 2};
    std::cout << "[2,1,1,2] -> " << rob(a3) << "\n";
}

File: HouseRobber.py
"""
Basic DSA - House Robber
You cannot rob two adjacent houses; maximize the total money robbed.
"""


def rob(nums):
    """
    Dynamic programming with O(1) state: (skip_prev_best, take_prev_best) rolled into two vars.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    skip = take = 0
    for x in nums:
        skip, take = take, max(take, skip + x)
    return max(skip, take)


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: House Robber")
    print("=" * 60)

    tests = [
        ([1, 2, 3, 1], 4),
        ([2, 7, 9, 3, 1], 12),
        ([2, 1, 1, 2], 4),
        ([5], 5),
        ([], 0),
    ]

    for nums, expected in tests:
        got = rob(nums)
        print(f"nums={nums} -> {got} (expected {expected})")